src/SqlBinding/SqlBindingUtilities.cs

```diff
@@ -162,31 +164,47 @@ public static SqlCommand BuildCommand(SqlAttribute attribute, SqlConnection conn
         }
 
         /// <summary>
-        /// Returns schema and tableName with quotes around them.
-        /// If there is no schema in fullName, SCHEMA_NAME is returned as schema.
+        /// Use ScriptDom to parse schema and tableName and return them with quotes.
+        /// If there is no schema in fullName, SCHEMA_NAME() is returned as schema.
         /// </summary>
         /// <param name="fullName">
         /// Full name of table, including schema (if exists).
         /// </param>
         public static void GetTableAndSchema(string fullName, out string quotedSchema, out string quotedTableName)
         {
-            // ensure names are properly escaped
-            string escapedFullName = fullName.Replace("'", "''");
+            var parser = new TSql150Parser(false);
+            var stringReader = new StringReader(fullName);
+            SchemaObjectName tree = parser.ParseSchemaObjectName(stringReader, out IList<ParseError> errors);
 
-            // defaults
-            quotedTableName = $"'{escapedFullName}'";
-            quotedSchema = "SCHEMA_NAME()"; // default to user schema
+            if (errors.Count > 0)
+            {
+                string errorMessages = "Encountered error(s) while parsing schema and table name:\n";
+                foreach (ParseError err in errors)
+                {
+                    errorMessages += $"{err.Message}\n";
+                }
+                throw new InvalidOperationException(errorMessages);
+            }
 
-            // remove [ ] from name if necessary
-            string cleanName = escapedFullName.Replace("]", string.Empty).Replace("[", string.Empty);
+            var visitor = new QuotedTSqlFragmentVisitor();
+            tree.Accept(visitor);
+            quotedSchema = visitor.quotedSchema;
+            quotedTableName = visitor.quotedTableName;
+        }
 
-            // if in format schema.table, split into two parts for query
-            string[] pieces = cleanName.Split('.');
+        /// <summary>
+        /// Get the schema and table name with quotes from the SchemaObjectName.  
+        /// </summary>
+        private class QuotedTSqlFragmentVisitor : TSqlFragmentVisitor
+        {
+            internal string quotedSchema;
+            internal string quotedTableName;
 
-            if (pieces.Length == 2)
+            public override void Visit(SchemaObjectName node)
             {
-                quotedSchema = $"'{pieces[0]}'";
-                quotedTableName = $"'{pieces[1]}'";
+                // ensure names with quotes are properly escaped
+                this.quotedSchema = node.SchemaIdentifier != null ? $"'{node.SchemaIdentifier.Value.Replace("'", "''")}'" : "SCHEMA_NAME()";
```

Comment:
Let's make a helper method in the utils to do the actual quoting to reduce chance of copy/paste errors (and we'll likely be wanting to re-use that in other contexts)

```
public static string QuoteIdentifier(string identifier)
{
    return $"'{identifier.Replace("'", "''")}'";
}
```